pkg/cloudprovider/aws/fleet/capacity.go

```diff
@@ -34,74 +31,48 @@ type Capacity struct {
 	launchTemplateProvider *LaunchTemplateProvider
 	subnetProvider         *SubnetProvider
 	nodeFactory            *NodeFactory
+	instanceProvider       *InstanceProvider
 }
 
 // Create a set of nodes given the constraints.
 func (c *Capacity) Create(ctx context.Context, constraints *cloudprovider.CapacityConstraints) (cloudprovider.CapacityPacking, error) {
-	// 1. Select a zone
-	zone, err := c.selectZone(ctx, constraints)
-	if err != nil {
-		return nil, fmt.Errorf("getting zone, %w", err)
-	}
-
-	// 2. Select a subnet, limited to selected zone.
-	subnet, err := c.selectSubnet(ctx, zone, constraints)
-	if err != nil {
-		return nil, fmt.Errorf("getting subnet, %w", err)
-	}
-
-	// 3. Detect launch template.
+	// 1. Compute Constraints
 	launchTemplate, err := c.launchTemplateProvider.Get(ctx, c.spec.Cluster)
 	if err != nil {
 		return nil, fmt.Errorf("getting launch template, %w", err)
 	}
-
-	// 3. Create Fleet.
-	createFleetOutput, err := c.ec2.CreateFleetWithContext(ctx, &ec2.CreateFleetInput{
-		Type: aws.String(ec2.FleetTypeInstant),
-		TargetCapacitySpecification: &ec2.TargetCapacitySpecificationRequest{
-			DefaultTargetCapacityType: aws.String(ec2.DefaultTargetCapacityTypeOnDemand), // TODO support SPOT
-			TotalTargetCapacity:       aws.Int64(1),                                      // TODO construct this more intelligently
-		},
-		LaunchTemplateConfigs: []*ec2.FleetLaunchTemplateConfigRequest{{
-			LaunchTemplateSpecification: &ec2.FleetLaunchTemplateSpecificationRequest{
-				LaunchTemplateName: launchTemplate.LaunchTemplateName,
-				Version:            aws.String("$Default"),
-			},
-			Overrides: []*ec2.FleetLaunchTemplateOverridesRequest{{
-				AvailabilityZone: aws.String(zone),
-				InstanceType:     aws.String("m5.large"), // TODO construct this more intelligently
-				SubnetId:         subnet.SubnetId,
-			}},
-		}},
-	})
+	instancePackings, err := c.instanceProvider.GetPackings(ctx, constraints.Pods, constraints.Overhead)
```

Comment:
After thinking a little more about the change, I strongly feel GetPackings should be in a separate package to calculate the instance options and make packing decisions. This package should only be making packing decisions and let someone else call create fleet API. As this thing itself will become complex and this package can be the brain that decides packaging and does this one thing really well.

Reply:
huge +1